src/MonitorProfiler/Utilities/TypeNameUtilities.cpp

```diff
@@ -0,0 +1,249 @@
+
+    tstring placeholderName;
+
+    IfFailRet(_profilerInfo->GetClassIDInfo2(classId,
+        &modId,
+        &classToken,
+        nullptr,
+        32,
+        &nTypeArgs,
+        typeArgs));
+
+    if (CORPROF_E_CLASSID_IS_ARRAY == hr)
+    {
+        flags = ClassFlags::Array;
+    }
+    else if (CORPROF_E_CLASSID_IS_COMPOSITE == hr)
+    {
+        // We have a composite class
+        flags = ClassFlags::Composite;
+    }
+    else if (CORPROF_E_DATAINCOMPLETE == hr)
+    {
+        // type-loading is not yet complete. Cannot do anything about it.
+        flags = ClassFlags::IncompleteData;
+    }
+    else if (FAILED(hr))
+    {
+        flags = ClassFlags::Error;
+    }
+
+    if (flags == ClassFlags::None)
+    {
+        IfFailRet(GetTypeDefName(nameCache, modId, classToken));
+
+        for (ULONG32 i = 0; i < nTypeArgs; i++)
+        {
+            if (typeArgs[i] != 0)
+            {
+                IfFailRet(GetClassInfo(nameCache, typeArgs[i]));
+            }
+        }
+    }
+
+    nameCache.AddClassData(modId, classId, classToken, flags, typeArgs, nTypeArgs);
+
+    return S_OK;
+}
+
+HRESULT TypeNameUtilities::GetTypeDefName(NameCache& nameCache, ModuleID moduleId, mdTypeDef classToken)
+{
+    HRESULT hr;
+    ComPtr<IMetaDataImport2> pMDImport;
+    IfFailRet(_profilerInfo->GetModuleMetaData(moduleId,
+        (ofRead | ofWrite),
+        IID_IMetaDataImport2,
+        (IUnknown**)&pMDImport));
+
+    mdToken tokenToProcess = classToken;
+    while (tokenToProcess != mdTokenNil)
+    {
+        std::shared_ptr<TokenData> tokenData;
+        if (nameCache.GetTokenData(moduleId, tokenToProcess, tokenData))
+        {
+            //We already processed this type (and therefore all of its outer classes)
+            break;
+        }
+
+        WCHAR wName[256];
+        DWORD dwTypeDefFlags = 0;
+        IfFailRet(pMDImport->GetTypeDefProps(tokenToProcess,
+            wName,
+            256,
+            NULL,
+            &dwTypeDefFlags,
+            NULL));
+
+        mdTypeDef outerTokenType = mdTokenNil;
+        if (IsTdNested(dwTypeDefFlags))
+        {
+            IfFailRet(pMDImport->GetNestedClassProps(tokenToProcess, &outerTokenType));
+        }
+        nameCache.AddTokenData(moduleId, tokenToProcess, outerTokenType, tstring(wName));
+        tokenToProcess = outerTokenType;
+    }
+
+    return S_OK;
+}
+
+HRESULT TypeNameUtilities::GetModuleInfo(NameCache& nameCache, ModuleID moduleId)
+{
+    if (moduleId == 0)
+    {
+        return E_INVALIDARG;
+    }
+
+    HRESULT hr;
+
+    std::shared_ptr<ModuleData> mod;
+    if (nameCache.GetModuleData(moduleId, mod))
+    {
+        return S_OK;
+    }
+
+    WCHAR moduleFullName[256];
+    ULONG nameLength = 0;
+    AssemblyID assemID;
+
+    IfFailRet(_profilerInfo->GetModuleInfo(moduleId,
+        nullptr,
+        256,
+        &nameLength,
+        moduleFullName,
+        &assemID));
+
+    WCHAR* ptr = nullptr;
+
+    int index = nameLength - 1;
```

Comment:
nit: Given that we're evaluating it down below (as opposed to solely using it for the purposes of iteration), can we give this a more descriptive name?